### PR TITLE
Promise and stub using done function, async await and assert fail (throw err)

### PR DESCRIPTION
**Pull request by DongThin**
- [ ] DONE  

```
test('This function completes successfully', function(done) {
    functionToTest (88_888_888, 5, jun23)
        .then((result) => {
            //assert here
            done();
        })
        .catch((function(error)  {
        //...
        done(); //Just call 'done' in case of error, it will stop the test running and display the error
});
})
```
- [ ]  ASYNC AWAIT
```
async function() {
  const  actual = await promiseFunction();
  console.log(response);
}
```

### CODE REVIEW

File: src/insurancesCalculator.js [hunk 1]
```diff
@@ -35,51 +35,55 @@ const POLICY_UPDATES = [
...
-        throw new Error("Invalid region entered. Please enter again! (1, 2, 3, 4)");
-    }
+        const grossBig = new Big(gross);
 
-    const policyUpdate = POLICY_UPDATES.find(function (update) {
-        return update.startDate.getTime() <= date.getTime()
-    })
+        if (!REGIONS.includes(region)) {
+            reject(new Error("Invalid region entered. Please enter again! (1, 2, 3, 4)"))
+        }
```
Comment (alang-dev) [inline src/insurancesCalculator.js:48]: will the program continue to run after reject?

File: test/insurancesCalculator.test.js [hunk 2]
```diff
@@ -170,20 +170,34 @@ describe('Calculate Insurances', function () {
...
         })
 
-        test('Throw Exception when entering invalid region', function () {
-            assert.throws(function () {
-                insurancesCalculator(93_678_123, 10, jun23);
-            }, Error, 'Invalid region entered. Please enter again! (1, 2, 3, 4)');
+        test('Throw Exception when entering invalid region', async function () {
+            try {
+                const actualInsurances = await insurancesCalculator(88_888_888, 5, jun23);
+                assert.fail('Expected insurancesCalculator to throw an error');
```
Comment (alang-dev) [inline test/insurancesCalculator.test.js:176]: why `assert.fail` here? and why try-cach in test?

File: test/insurancesCalculator.test.js [hunk 3]
```diff
@@ -340,19 +354,29 @@ describe('Calculate Insurances', function () {
...
         })
 
-        test('Throw Exception when entering invalid region, after 1/7/2023', function () {
-            assert.throws(function () {
-                insurancesCalculator(93_678_123, 0, 10);
-            }, Error, 'Invalid region entered. Please enter again! (1, 2, 3, 4)');
+        test('Throw Exception when entering invalid region, after 1/7/2023', async function () {
+            try {
+                const actualInsurances = insurancesCalculator(93_678_123, 0, 10);
+                assert.fail("Invalid region entered. Please enter again! (1, 2, 3, 4)");
```
Comment (alang-dev) [inline test/insurancesCalculator.test.js:360]: why?

File: test/net.test.js [hunk 4]
```diff
@@ -38,14 +38,18 @@ describe("Calculate net salary", function () {
                 totalTax: 0,
                 netSalary: 10999997.5
             };
-            assert.deepStrictEqual(actualPayslip, expectedPayslip);
+            net(12_290_500, 0).then(function (result) {
+                assert.deepStrictEqual(actualPayslip, expectedPayslip);
```
Comment (alang-dev) [inline test/net.test.js:42]: How is it even possible? The `actualPayslip` havenot declared yet1

**Review by alang-dev — CHANGES_REQUESTED** [review]

File: test/net.test.js [hunk 5]
```diff
@@ -5,79 +5,91 @@ const sinon = require('sinon');
...
+
+            net(gross, 1).then(function (result) {
+
+                assert.deepStrictEqual(result, expectedPayslip);
+                done();
+
+            }).catch(function () {
+                done()
+            })
+
```
Comment (alang-dev) [inline test/net.test.js:87]: we have to verify tha calcTax with arg = 0. Pls apply next test case as well

File: test/net.test.js [hunk 6]
```diff
@@ -5,79 +5,91 @@ const sinon = require('sinon');
 const proxyquire = require("proxyquire");
 
 describe("Calculate net salary", function () {
+    let insurancesCalcStub;
```
Comment (alang-dev) [inline test/net.test.js:8]: line 1 is no longer use.

File: test/net.test.js [hunk 7]
```diff
@@ -140,45 +157,34 @@ describe("Calculate net salary", function () {
...
+    describe('Check Insurance & Taxes deduction', function (done) {
 
-        afterEach(function () {
-            sinon.restore();
-        })
 
-        test('It should correctly deduct total Insurance and total tax', function () {
+        test('It should correctly deduct total Insurance and total tax', async function () {
             // Arrange
             mockedInsurance = { total: 1, insurances: [] };
```
Comment (alang-dev) [inline test/net.test.js:177]: const?

**Review by alang-dev — CHANGES_REQUESTED** [review]

File: test/net.test.js [hunk 8]
```diff
@@ -1,184 +1,157 @@
...
-        test('No dependent deducted', function () {
+        afterEach(function () {
+            sinon.restore();
+        })
+
+        test('No dependent deducted', async function () {
             //Gross income below 11 million is tax-exempt.
-            const actualPayslip = net(12_290_500, 0);
+            const gross = 12_290_500;
+            taxCalcStub.withArgs(0).resolves(mockedTaxes);
```
Comment (alang-dev) [inline test/net.test.js:48]: taxCalcStub is stubbed in beforeEach block. We can use that stub.

In this test case, we need to verify that the stub is called with expected args. Something like:
`assert(callback.calledWith('arg1', 'arg2'))`

Check the rest of the test cases if we need to verify the tax stub and insurance stub

File: test/taxCalculator.test.js [hunk 9]
```diff
@@ -1,57 +1,87 @@
...
             rates: [
                 { name: 'Tax level 1', rate: 0.05, amount: 250_000 },
                 { name: 'Tax level 2', rate: 0.1, amount: 500000 },
                 { name: 'Tax level 3', rate: 0.15, amount: 300000 }],
             totalTax: 1050000
         };
-        assert.deepStrictEqual(actualTaxes, expectedTaxes);
+
+        await calculateTaxes(12_000_000).then(function (taxes) {
+
```
Comment (alang-dev) [inline test/taxCalculator.test.js:74]: why do we have so many redundant spaces here?

File: test/taxCalculator.test.js [hunk 10]
```diff
@@ -1,57 +1,87 @@
 const calculateTaxes = require('../src/taxCalculator.js')
+const Big = require('big.js');
 
 const assert = require('assert');
 const test = require('mocha').it;
 
 describe('Calculate taxes', function () {
 
-    test('It applies tax rate 1 only', function () {
+    test('It should return taxes correctly when taxableIncome is less than or equal to zero', async function () {
```
Comment (alang-dev) [inline test/taxCalculator.test.js:9]: pls find shorter name. `It should return no tax when ...`

**Review by alang-dev — CHANGES_REQUESTED** [review]

File: test/taxCalculator.test.js (hunk 9 above)
Comment (DongThin) [reply test/taxCalculator.test.js:74]: just 1 line

File: test/net.test.js [hunk 11]
```diff
@@ -1,184 +1,159 @@
...
             };
-            assert.deepStrictEqual(actualPayslip, expectedPayslip);
-        })
 
-        test('Dependent with no taxable income', function () {
-            //afterInsurance below 11 million is tax-exempt.
+            await net(gross).then(function (result) {
 
-            const actualPayslip = net(12_290_500, 1);
+                assert(taxCalcStub.calledWith(1290500));
```
Comment (alang-dev) [inline test/net.test.js:64]: move it outside then (after await). The reasons:
- net.then is a function to get the actual payslip to continue next statements.
- taxCalcStub has nothing to do inside then.

File: test/taxCalculator.test.js [hunk 12]
```diff
@@ -114,14 +148,16 @@ describe('Calculate taxes', function () {
...
     })
 
-    test('Total tax is equal total of tax per rate ', function () {
+    test('Total tax is equal total of tax per rate ', async function () {
         //tax.totalTax = tax.totalTax.plus(taxPaidForEachRate);
 
-        const actualTaxes = calculateTaxes(29_000_000);
-        assert.deepStrictEqual(actualTaxes.totalTax, 4_150_000);
+        await calculateTaxes(29_000_000).then(function (taxes) {
+            assert.deepStrictEqual(taxes.totalTax, 4_150_000)
```
Comment (alang-dev) [inline test/taxCalculator.test.js:160]: should calc sum of paid tax per rates from taxes And then assert that sum with taxes.totalTax

File: test/taxCalculator.test.js [hunk 13]
```diff
@@ -1,57 +1,84 @@
...
 describe('Calculate taxes', function () {
 
-    test('It applies tax rate 1 only', function () {
+    test('It should return no tax when taxableIncome is lte 0', async function () {
+        // taxableIncome <= 5m (tax level 1)
+        const taxableIncome = -1;
+
+        const expectedTaxes = {
+            rates: [],
+            totalTax: new Big(0)
```
Comment (alang-dev) [inline test/taxCalculator.test.js:15]: why Big? The return type must be number to make it consistent

**Review by alang-dev — CHANGES_REQUESTED** [review]

File: src/insurancesCalculator.js [hunk 14]
```diff
@@ -35,51 +35,56 @@ const POLICY_UPDATES = [
  * @param {Number} gross 
  * @param {1|2|3|4} region 
  * @param {Date} date 
- * @returns { {
- * total: number,
- * details: [{name: string, amount: number}]
-* }}
+ * @returns {Promise<{total: number, details: { name: string, amount: number }[]}>}
```
Comment (alang-dev) [inline src/insurancesCalculator.js:38]: Wrong property name. It should be `insurances` instead of  `details`. See the return obj in last line

**Review by alang-dev — CHANGES_REQUESTED** [review]

**Review by alang-dev — APPROVED** [review]

**Pull request merged**